src/server/pfs/server/driver.go

```diff
@@ -264,23 +265,52 @@ func (d *driver) getPermissions(ctx context.Context, repo *pfs.Repo) ([]auth.Per
 func (d *driver) listRepo(ctx context.Context, includeAuth bool, repoType string, projects []*pfs.Project, cb func(*pfs.RepoInfo) error) error {
 	authSeemsActive := true
 	repoInfo := &pfs.RepoInfo{}
+
+	// Helper func to filter out repos based on projects.
 	projectsFilter := make(map[string]bool)
 	for _, project := range projects {
-		projectsFilter[project.String()] = true
+		projectsFilter[project.GetName()] = true
+	}
+	keep := func(repo *pfs.Repo) bool {
+		// Assume the user meant all projects by not providing any projects to filter on.
+		if len(projectsFilter) == 0 {
+			return true
+		}
+		return projectsFilter[repo.Project.GetName()]
+	}
+
+	// Helper func to check whether a user is allowed to see the given repo in the result.
+	// Cache the project level access because it applies to every repo within the same project.
+	checkProjectAccess := miscutil.CacheFunc(func(project string) error {
+		return d.env.AuthServer.CheckProjectIsAuthorized(ctx, &pfs.Project{Name: project}, auth.Permission_PROJECT_LIST_REPO)
+	}, 100 /* size */)
+	checkAccess := func(ctx context.Context, repo *pfs.Repo) error {
+		if err := checkProjectAccess(repo.Project.GetName()); err != nil {
+			if !errors.As(err, &auth.ErrNotAuthorized{}) {
+				return err
+			}
+			// Allow access if user has the right permissions at the individual Repo-level.
+			return d.env.AuthServer.CheckRepoIsAuthorized(ctx, repo, auth.Permission_REPO_READ)
+		}
+		return nil
 	}
 
 	processFunc := func(string) error {
-		// Assume the user meant all projects by not providing any projects to filter on.
-		if len(projectsFilter) > 0 && !projectsFilter[repoInfo.Repo.Project.String()] {
+		if !keep(repoInfo.Repo) {
```

Comment:
I wanted to put the exclude case on top so that we don't wrap majority of the actual logic in an if-block.

In terms of semantics, here `keep()` returns true if the repo's project satisfies the user's text-based filter via `--project`, otherwise we can return early to avoid unnecessary auth calls since we know the user doesn't want to see this repo at all. Similarly, `checkAccess()` is returns an error if the user doesn't have access to this repo. Returning `nil` is our way of skipping this repo since it won't be returned to the client.

Reply:
ahh totally misread the keep() semantics 👍